Snap 1499 2 : Support for snapshot write operation

suranjan opened the pull request:
## Changes proposed in this pull request
* Implemented Rollback and write confict in case of snapshot mode. In case of rollback, put the entry back into ABRM. Also record the events so that no exception list is maintained.
* Changes needed for smart connector. Added some procedure required for snapshot tx write operation from snappy.
* We use the txId to either start/masqueradeAs/commit/rollback a transaction on a given connection.

## Patch testing

precheckin and tet

## ReleaseNotes changes

(Does this change require an entry in ReleaseNotes? If yes, has it been added to it?)

## Other PRs 

(Does this change require changes in other projects- snappydata, spark, spark-jobserver, aqp? Add the links of PR of the other subprojects that are related to this change)

Commit:
Changes needed for smart connector

Commit:
Merge branch 'snappy/master' into SNAP-1499-2

Commit:
Changed for Rollback and write confict in case of snapshot mode

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractRegionMap.java:
```diff
@@ -3980,20 +3975,21 @@ public RegionEntry basicPut(EntryEventImpl event,
                       final int oldSize = event.getLocalRegion().calculateRegionEntryValueSize(re);
-                      if (shouldCopyOldEntry(owner, event) && re.getVersionStamp() != null && re
-                          .getVersionStamp().asVersionTag().getEntryVersion() > 0) {
+                      if (shouldCopyOldEntry(owner, event) && re.getVersionStamp() != null ) {
```
Just got interested in shouldCopyOldEntry method, though changes not in this PR. It looks like this can just be a final field and we don't need method invocation all the time. Isn't it?

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractRegionMap.java:
```diff
@@ -4144,6 +4140,24 @@ else if (re != null && owner.isUsedForPartitionedRegionBucket()) {
+      TXState localState = event.getTXState().getLocalTXState();
+      if(!firstEntry(re)) {
+        if (!localState.checkEntryVersion(event.getRegion(), re)) {
+          throw new ConflictException("The value has been changed.");
```
I think "has changed" is good. Remove "been" from the message.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/AbstractRegionMap.java:
```diff
@@ -4144,6 +4140,24 @@ else if (re != null && owner.isUsedForPartitionedRegionBucket()) {
+    if (event.getTXState() != null && event.getTXState().isSnapshot()) {
+      TXState localState = event.getTXState().getLocalTXState();
+      if(!firstEntry(re)) {
+        if (!localState.checkEntryVersion(event.getRegion(), re)) {
```
Should we change the name of checkEntryVersion to something like checkEntryInSnapshot? I had to look at the method to understand what kind of check is this.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/NonLocalRegionEntry.java:
```diff
@@ -46,7 +47,7 @@
 import com.gemstone.gemfire.internal.shared.Version;
 
-public class NonLocalRegionEntry implements RegionEntry, VersionStamp {
+public class NonLocalRegionEntry implements RegionEntry, VersionStamp, DataSerializable {
```
Why implementing DataSerializable required? It already implements DataSerializableFixedID no? But I saw the code and in that it is not registered as a fixed id class in DSFIDFactory but then why does it implement toData and fromData? Could not understand.
And I guess DataSerializable is more costly than fixedId thing. It writes the full classname during serialization? Please check if fixed id will be better.

kneeraj replied:
Looks like, as discussed, serialization of this entry not required as this will never be persisted or sent over the wire.

suranjan replied:
Yes will remove as discussed. It was for persisting NonLocalRegionEntry which was part of other design.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -852,18 +857,50 @@ public void rollback(final Object callbackArg) {
+        }
+        cleanEx = processCleanupException(t, cleanEx);
+      } finally {
+
```
If finally block is empty why do we need it? Remove it. I think the compiler will remove it. But anyways.
This code and the below can be put in a single method. Exact same code.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -1159,6 +1160,49 @@ public boolean execute(Object key) {
+    GemFireCacheImpl cache = GemFireCacheImpl.getExisting();
+    final LogWriterI18n logger = getTxMgr().getLogger();
+    // No need to check for snapshot if we want to enable it for RC.
+    if (cache.snapshotEnabled()) {
```
This check can be called before calling publishRecordedVersions. If snapshot is not enabled then this can be avoided. There are a lot of method calls like isSnapshot, snapShotEnabled then there is a flag called DEFAULT_SNAPSHOT_ENABLED_TX which I feel are doing duplicate kind of work of checking snapshot enabled etc with little differences. At some point we should do this cleanup especially for row tables it will become important for low latency.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -1159,6 +1160,49 @@ public boolean execute(Object key) {
+      if (isSnapshot() || cache.snapshotEnabledForTX()) {
+        // first take a lock at cache level so that we don't go into deadlock or sort array before
+        // This is for tx RC, for snapshot just record all the versions from the queue
+        cache.acquireWriteLockOnSnapshotRvv();
```
It is better to acquire the lock inside try. Here it does not matter but if someone refactors a bit and there is a scope of failure between acquire and try block start then the write lock won't be released.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -1159,6 +1160,49 @@ public boolean execute(Object key) {
+        // tx may not record version in snapshot so non tx reads while taking
+        // snapshot may miss it. in case of commit just copy the rvv to snapshot so that
+        // any future non tx read will get all the entries
+        cache.acquireWriteLockOnSnapshotRvv();
```
Same as above

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -1347,6 +1391,101 @@ final void cleanupTXRS(final TXRegionState[] regionStates,
+    Iterator<Object> itr2 = committedEntryReference.iterator();
+    Iterator<LocalRegion> itr3 = regionReference.iterator();
+
+    GemFireCacheImpl.getInstance().getLogger().info("Rolling back the changes.. " + unCommittedEntryReference.size()
```
You want to keep this in info. Specially since conflict is added, this might fill the logs. No?

suranjan replied:
Will do all log related cleanup.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -1347,6 +1391,101 @@ final void cleanupTXRS(final TXRegionState[] regionStates,
+      // so we will have to store the version of uncommitted too?
+      // We can get it from version that we store in TXState. We need to map RE with the version
+      // What if multiple updates on same RE.
+      GemFireCacheImpl.getInstance().getLogger().info("SKKS rolling back the changes.. ");
```
SKSK...

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -3880,6 +4027,54 @@ private Object getOldVersionedEntry(LocalRegion dataRegion, Object key, RegionEn
+            Thread.sleep(100 * numtimes);
+          }
+        } catch (InterruptedException e) {
+          e.printStackTrace();
```
You wan't to log it and fail instead?

kneeraj replied:
Same as above.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java:
```diff
@@ -3880,6 +4027,54 @@ private Object getOldVersionedEntry(LocalRegion dataRegion, Object key, RegionEn
+        try {
+          //TODO: Suranjan Should we wait indefinitely? or throw warning and return the current entry.
+          // As
+          if (numtimes < 10) {
```
Wow. This might cause unnecessary delay. When getOldVersionedEntry method is called, i.e. this method, the caller is sure that it will find it in the oldEntryMap. Can there be a situation that this unnecessarily keeps on waiting. Second thing is that 10 ms looks too long for waiting for a local op to populate the map. May be we can start with couple and increase to 10 later ( instead of 100 later )

suranjan replied:
Will throw exception here after some wait.

kneeraj commented on gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXStateProxy.java:
```diff
@@ -177,7 +177,7 @@ public final TXStateProxy newTXStateProxy(final TXManagerImpl txMgr,
 
   /** True if operations related to transactions have to be logged globally. */
-  private static boolean VERBOSE = VERBOSE_ON();
+  private static boolean VERBOSE = true;//VERBOSE_ON();
```
Revert.

kneeraj reviewed (commented):
Mostly done. A couple of files and test left. Will do in a bit.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -2471,7 +2473,7 @@ public static void COMMIT_SNAPSHOT_TXID(String txId) throws SQLException, Standa
       if (GemFireXDUtils.TraceExecution) {
         SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_EXECUTION,
-            "in procedure COMMIT_SNAPSHOT_TXID() " + txId + " for connid " + tc.getConnectionID()
+            "in procedure COMMIT_SNAPSHOT_TXID() SURANJAN " + txId + " for connid " + tc.getConnectionID()
```
Suranjan

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -2494,7 +2496,46 @@ public static void COMMIT_SNAPSHOT_TXID(String txId) throws SQLException, Standa
     if (GemFireXDUtils.TraceExecution) {
       SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_EXECUTION,
-          "in procedure COMMIT_SNAPSHOT_TXID() afer commit" + txId + " for connid " + tc.getConnectionID()
+          "in procedure COMMIT_SNAPSHOT_TXID() SURANJAN afer commit" + txId + " for connid " + tc.getConnectionID()
```
remove your name from all logs.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -2462,6 +2463,7 @@ public static void SET_NANOTIMER_TYPE(Boolean useNativeTimer, String nativeTimer
         GfxdSystemProcedureMessage.SysProcMethod.setNanoTimerType, false, true);
   }
 
+  // TODO: send to all nodes after doing local commit
```
Is this still TODO? If this is required, how this is working without this. Same question for rollback as well.

suranjan replied:
Not required as we are now using same connection once transaction has begun.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/NonLocalRowLocationRegionEntry.java:
```diff
@@ -314,15 +314,17 @@ public InputStream getStream() throws StandardException {
   @Override
   public DataValueDescriptor getClone() {
-    throw new UnsupportedOperationException("unexpected invocation");
+    return this;
```
Why was this change required?

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/store/NonLocalRowLocationRegionEntry.java:
```diff
@@ -314,15 +314,17 @@ public InputStream getStream() throws StandardException {
   @Override
   public DataValueDescriptor recycle() {
-    throw new UnsupportedOperationException("unexpected invocation");
+    return this;
```
Same question as above.

kneeraj commented on gemfirexd/tools/src/test/java/com/pivotal/gemfirexd/jdbc/transactions/snapshot/SnapshotTransactionTest.java:
```diff
@@ -100,6 +94,120 @@ public void testRVVSnapshotContains() throws Exception {
+    while (rs.next()) {
+      numRows++;
+    }
+    assertEquals("ResultSet should contain two rows ", 1, numRows);
```
You mean 1  row I guess...

kneeraj replied:
Also PK violation should have come. What am I missing here?

kneeraj replied:
What exactly are we testing in this test?

suranjan replied:
Yes, it comes. This test is still under development was testing autocommit behavior and rollback.

kneeraj commented on gemfirexd/tools/src/test/java/com/pivotal/gemfirexd/jdbc/transactions/snapshot/SnapshotTransactionTest.java:
```diff
@@ -100,6 +94,120 @@ public void testRVVSnapshotContains() throws Exception {
+    while (rs.next()) {
+      numRows++;
+    }
+    assertEquals("ResultSet should contain two rows ", 1, numRows);
```
Again instead of two in the message you must be meaning one. BTW I am missing something basic. The name of the test has RC in it and I don't see that level ever being set anywhere.

kneeraj commented on gemfirexd/tools/src/test/java/com/pivotal/gemfirexd/jdbc/transactions/snapshot/SnapshotTransactionTest.java:
```diff
@@ -100,6 +94,120 @@ public void testRVVSnapshotContains() throws Exception {
+    while (rs.next()) {
+      numRows++;
+    }
+    assertEquals("ResultSet should contain two rows ", 1, numRows);
```
two?

suranjan replied:
Will correct messages.

kneeraj commented on gemfirexd/tools/src/test/java/com/pivotal/gemfirexd/jdbc/transactions/snapshot/SnapshotTransactionTest.java:
```diff
@@ -1011,6 +1119,416 @@ public void run() {
+    Statement st = conn.createStatement();
+    ResultSet rs;
+    int numRows = 0;
+    st.execute("Create table t1 (c1 int not null , c2 int not null, primary key(c1)) replicate persistent enable concurrency checks " + getSuffix());
```
All your test is using replicate persistent. Partiton persistent should also be added. Isn't it?

Commit:
Code review changes

  Cleaned up code
  Fixed some issue with autocommit of jdbc insert in column table
  Log cleanup
  Fixed an issue where a txState was dangling in currentRunningTxState in smart connector mode

Commit:
Merge branch 'snappy/master' into SNAP-1499-2

Conflicts:
	gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/TXState.java

Commit:
Removed dead code

Commit:
Fixed a test

Commit:
Fixed a test issue

  This could cause some test in store checkin to fail.
  Specially test related to TransactionRRDunit
